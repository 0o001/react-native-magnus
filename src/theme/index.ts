export { ThemeContext } from './themeContext';
export { withTheme } from './withTheme';
export { ThemeProvider, ThemeProviderProps } from './themeProvider.component';

export {
  ThemeType,
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from './type';
