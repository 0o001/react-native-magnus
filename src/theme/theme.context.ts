import * as React from 'react';

import { ThemeType } from './type';

const defaultValue: ThemeType = {};

export interface ThemeContextType {
  theme: ThemeType;
  setTheme?: (theme?: ThemeType) => void;
}

export const THEME_CONTEXT_DEFAULT_VALUE = {
  theme: defaultValue,
};

export const ThemeContext: React.Context<ThemeContextType> = React.createContext(
  THEME_CONTEXT_DEFAULT_VALUE
);
