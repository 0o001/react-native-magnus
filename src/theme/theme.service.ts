import { Dimensions } from 'react-native';

import { ThemeType, BreakpointType } from './type';
import { defaultTheme } from '../style';

/**
 * merge user theme with default theme
 *
 * @param customTheme
 *
 * @return computedTheme computed theme variables object
 */
export function computeTheme(customTheme: ThemeType | undefined): ThemeType {
  const computedTheme = { ...defaultTheme };

  if (!customTheme) {
    return computedTheme;
  }

  if (customTheme.colors) {
    computedTheme.colors = {
      ...defaultTheme.colors,
      ...customTheme.colors,
    };
  }

  if (customTheme.fontSize) {
    computedTheme.fontSize = {
      ...defaultTheme.fontSize,
      ...customTheme.fontSize,
    };
  }

  if (customTheme.borderRadius) {
    computedTheme.borderRadius = {
      ...defaultTheme.borderRadius,
      ...customTheme.borderRadius,
    };
  }

  if (customTheme.name) {
    computedTheme.name = customTheme.name;
  }

  if (customTheme.breakpoints) {
    computedTheme.breakpoints = customTheme.breakpoints;
    computedTheme.breakpoint = calculateCurrentDeviceBreakpoint(
      customTheme.breakpoints
    );
  }

  return computedTheme;
}

/**
 * m="10"
 *
 * @param value
 * @param theme
 */
export const createSpacingStyles = (props: any, theme: any) => {
  let propKeys: any;

  propKeys = {
    p: 'padding',
    pt: 'paddingTop',
    pr: 'paddingRight',
    pb: 'paddingBottom',
    pl: 'paddingLeft',
    px: 'paddingHorizontal',
    py: 'paddingVertical',
    m: 'margin',
    mt: 'marginTop',
    mr: 'marginRight',
    mb: 'marginBottom',
    ml: 'marginLeft',
    mx: 'marginHorizontal',
    my: 'marginVertical',
    ms: 'marginStart',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      computedStyle[styleProperty] = getThemeProperty(
        theme,
        props[propKey],
        props[propKey]
      );
    }
  });

  return computedStyle;
};

/**
 * create directional style for properties like
 * top, left, right, bottom, horizontal, vertical
 *
 * @param key
 * @param value
 */
export const createDirectionalStyles = (
  key: string,
  value: any,
  type: string = 'string'
) => {
  if (
    typeof (value.x || value.y || value.b || value.t || value.l || value.r) ===
    'undefined'
  ) {
    return createStyle(`${key}`, value, type);
  }

  let computedStyle = {};

  if (typeof value.x !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Horizontal`, value.x, type),
    };
  }

  if (typeof value.y !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Vertical`, value.y, type),
    };
  }

  if (typeof value.t !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Top`, value.t, type),
    };
  }

  if (typeof value.r !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Right`, value.r, type),
    };
  }

  if (typeof value.b !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Bottom`, value.b, type),
    };
  }

  if (typeof value.l !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Left`, value.l, type),
    };
  }

  return computedStyle;
};

/**
 * extract the theme property from theme

 * @param theme
 * @param value
 */
export const getThemeProperty = (theme: any, value: any, fallback: any) => {
  if (typeof theme[value] !== 'undefined') {
    return theme[value];
  }

  return fallback;
};

/**
 * borderRadius="xl"
 *
 * @param value
 * @param theme
 */
export const createBorderRadiusStyles = (props: any, theme: any) => {
  let propKeys: any;

  propKeys = {
    rounded: 'borderRadius',
    roundedTopLeft: 'borderTopLeftRadius',
    roundedTopRight: 'borderTopRightRadius',
    roundedBottomLeft: 'borderBottomLeftRadius',
    roundedBottomRight: 'borderBottomRightRadius',
    roundedTop: ['borderTopLeftRadius', 'borderTopRightRadius'],
    roundedLeft: ['borderTopLeftRadius', 'borderBottomLeftRadius'],
    roundedRight: ['borderTopRightRadius', 'borderBottomRightRadius'],
    roundedBottom: ['borderBottomLeftRadius', 'borderBottomRightRadius'],
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      if (styleProperty instanceof Array) {
        styleProperty.forEach((property) => {
          computedStyle[property] = getThemeProperty(
            theme,
            props[propKey],
            props[propKey]
          );
        });
      } else {
        computedStyle[styleProperty] = getThemeProperty(
          theme,
          props[propKey],
          props[propKey]
        );
      }
    }
  });

  return computedStyle;
};

/**
 * adds flex property in styles
 *
 * @param props
 */
export const createFlexStyles = (props: any) => {
  let propKeys: any;

  propKeys = {
    flex: 'flex',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      computedStyle[styleProperty] = props[propKey];
    }
  });

  return computedStyle;
};

/**
 * create borderWidth styles
 *
 * @param value
 */
export const createBorderWidthStyles = (props: any) => {
  let propKeys: any;

  propKeys = {
    borderWidth: 'borderWidth',
    borderTopWidth: 'borderTopWidth',
    borderEndWidth: 'borderEndWidth',
    borderLeftWidth: 'borderLeftWidth',
    borderRightWidth: 'borderRightWidth',
    borderBottomWidth: 'borderBottomWidth',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      computedStyle[styleProperty] = props[propKey];
    }
  });

  return computedStyle;
};

/**
 * create borderRadius styles
 *
 * @param value
 */
export const createBorderColorStyles = (props: any, theme: any) => {
  let propKeys: any;

  propKeys = {
    borderColor: 'borderColor',
    borderTopColor: 'borderTopColor',
    borderRightColor: 'borderTopColor',
    borderLeftColor: 'borderLeftColor',
    borderBottomColor: 'borderBottomColor',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      if (styleProperty instanceof Array) {
        styleProperty.forEach((property) => {
          computedStyle[property] = getThemeProperty(
            theme,
            props[propKey],
            props[propKey]
          );
        });
      } else {
        computedStyle[styleProperty] = getThemeProperty(
          theme,
          props[propKey],
          props[propKey]
        );
      }
    }
  });

  return computedStyle;
};

/**
 * adds shadows property in styles
 *
 * @param props
 */
export const createShadowStyles = (props: any, theme: any) => {
  let computedStyle: any = {};

  if (props.shadow) {
    computedStyle = {
      ...theme.shadow[props.shadow],
      shadowColor: getThemeProperty(theme.colors, props.shadowColor, 'white'),
    };
  }

  return computedStyle;
};

/**
 * create style object
 *
 * @param key
 * @param value
 */
export const createStyle = (key: string, value: any, type: string) => {
  return {
    [key]: type === 'number' ? Number(value) : value,
  };
};

/**
 *
 * @param props
 * @param computedStyle
 * @param styleProperty
 * @param atomicProperty
 * @param type
 */
export const addPropToComputedStyle = (
  props: any,
  computedStyle: any,
  styleProperty: string,
  atomicProperty: string,
  type: string
) => {
  if (props[atomicProperty]) {
    computedStyle = {
      ...computedStyle,
      ...createStyle(styleProperty, props[atomicProperty], type),
    };
  }

  return computedStyle;
};

/**
 * position="absolute"
 * top={10}
 *
 * @param value
 */
export const createPositionStyle = (props: any) => {
  let propKeys: any;

  propKeys = {
    position: 'position',
    top: 'top',
    left: 'left',
    right: 'right',
    bottom: 'bottom',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      computedStyle[styleProperty] = props[propKey];
    }
  });

  return computedStyle;
};

/**
 * position="absolute"
 * top={10}
 * w={[100%, 50%]}
 *  1005
 *  50
 * const DEFAULT_BREAKPOINTS = {
  sm: 320,
  md: 480,
  lg: 768,
  xl: 1024,
};
 *
 * @param value
 */
export const createResponsiveStyle = (theme: ThemeType, value: any) => {
  if (typeof value === 'string' || typeof value === 'number') {
    return value;
  }

  if (!theme.breakpoints) {
    throw new Error('No breakpoints defined');
  }

  if (!theme.breakpoint) {
    throw new Error('Not able to calculate current breakpoint');
  }

  const currentBreakpointValue = theme.breakpoints[theme.breakpoint];

  // if prop is array like bg=["red500", "blue500"]
  if (Array.isArray(value)) {
    if (value.length === 1) return value[0];

    let currentValue = value[0];

    if (currentBreakpointValue) {
      value.forEach((propValue, index) => {
        if (
          currentBreakpointValue >=
          theme.breakpoints[Object.keys(theme.breakpoints)[index]]
        ) {
          currentValue = propValue;
        } else return;
      });
    }

    return currentValue;
  }

  // if prop is object like bg={{ sm: 'red500', lg: 'blue500'}}
  if (typeof value === 'object') {
    let currentValue = value[Object.keys(value)[0]];

    Object.keys(value).forEach((breakpointKey, index) => {
      if (currentBreakpointValue >= theme.breakpoints[breakpointKey]) {
        currentValue = value[breakpointKey];
      } else return;
    });

    return currentValue;
  }

  return value;
};

/**
 * calculates current device breakpoint
 *
 * @param breakpoints
 */
export const calculateCurrentDeviceBreakpoint = (
  breakpoints: BreakpointType,
  currentWidth: number = Dimensions.get('window').width
) => {
  let currentBreakpoint = Object.keys(breakpoints)[0];

  Object.keys(breakpoints).forEach((breakpointKey: string) => {
    const breakpointValue = breakpoints[breakpointKey];

    if (currentWidth > breakpointValue) {
      currentBreakpoint = breakpointKey;
    } else return;
  });

  return currentBreakpoint;
};
