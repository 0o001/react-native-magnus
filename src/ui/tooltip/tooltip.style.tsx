import { StyleSheet, Platform } from 'react-native';

import {
  getThemeProperty,
  createFlexStyles,
  createPositionStyle,
  createSpacingStyles,
  createBorderWidthStyles,
  createBorderColorStyles,
  createBorderRadiusStyles,
} from '../../theme/theme.service';

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any, state: any) => {
  const computedStyle: any = {};

  computedStyle.container = {
    flexDirection: props.flexDir,
    flexWrap: props.flexWrap,
    alignItems: props.alignItems,
    justifyContent: props.justifyContent,
    backgroundColor: getThemeProperty(theme.colors, props.bg, 'transparent'),
    ...createFlexStyles(props),
    ...createBorderWidthStyles(props),
    ...createSpacingStyles(props, theme.spacing),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius),
  };

  computedStyle.image = {
    ...createBorderRadiusStyles(props, theme.borderRadius),
  };

  if (props.shadow) {
    computedStyle.container = {
      ...computedStyle.container,
      ...theme.shadow[props.shadow],
      shadowColor: getThemeProperty(theme.colors, props.shadowColor, 'white'),
    };
  }

  if (props.h) {
    computedStyle.container = {
      ...computedStyle.container,
      height: props.h,
    };
  }

  if (props.w) {
    computedStyle.container = {
      ...computedStyle.container,
      width: props.w,
    };
  }

  if (props.minH) {
    computedStyle.container = {
      ...computedStyle.container,
      minHeight: props.minH,
    };
  }

  if (props.minW) {
    computedStyle.container = {
      ...computedStyle.container,
      minWidth: props.minW,
    };
  }

  computedStyle.triangle = {
    width: 0,
    height: 0,
    left: state.left + state.buttonWidth / 2 - 7.5,
    top: state.invert ? state.menuHeight : 0,
    borderStyle: 'solid',
    borderLeftWidth: 8,
    borderRightWidth: 8,
    borderBottomWidth: 10,
    borderLeftColor: 'transparent',
    borderRightColor: 'transparent',
    borderBottomColor: getThemeProperty(theme.colors, props.bg, 'transparent'),
  };

  computedStyle.text = {
    color: getThemeProperty(theme.colors, props.color, 'white'),
  };

  // merging style props to computed style
  if (props.style) {
    computedStyle.container = {
      ...computedStyle.container,
      ...props.style,
    };
  }

  return StyleSheet.create(computedStyle);
};
