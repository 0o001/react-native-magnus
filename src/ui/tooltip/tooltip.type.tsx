import { ViewProps as RNViewProps } from "react-native";

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
  ShadowPropsType
} from "../../theme";

export interface TooltipProps
  extends RNViewProps,
    BorderPropsType,
    SpacingPropsType,
    ShadowPropsType,
    RoundedPropsType {
  bg?: string;
  h?: number;
  w?: number;
  minW?: number;
  minH?: number;
  maxW?: number;
  maxH?: number;
  animationDuration?: number;
  onHidden?: () => void;
  opacity?: number;
  zIndex?: number;
  text: string | React.ReactNode;
  fontSize?: string;
  textDecorLine?:
    | "none"
    | "underline"
    | "line-through"
    | "underline line-through";
  textDecorStyle?: "solid" | "double" | "dotted" | "dashed";
  fontStyle?: "normal" | "italic";
  textDecorColor?: string;
  fontWeight?: string;
  lineHeight?: number;
  textAlign?: "auto" | "left" | "right" | "center" | "justify";
  textTransform?: "none" | "uppercase" | "lowercase" | "capitalize";
  letterSpacing?: number;
}

export interface TriangelProps extends RNViewProps {
  invert: boolean;
}
