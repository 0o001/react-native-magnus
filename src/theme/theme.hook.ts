import React, { useState } from "react";

import { ThemeType } from "./type";
import { computeTheme } from "./theme.service";
import { ThemeContextType } from "./theme.context";

export const useTheme = (initialTheme?: ThemeType): ThemeContextType => {
  const [theme, setTheme] = useState(computeTheme(initialTheme));

  const setCurrentTheme = React.useCallback((newTheme?: ThemeType): void => {
    setTheme(computeTheme(newTheme));
  }, []);

  return {
    theme,
    setTheme: setCurrentTheme
  };
};
