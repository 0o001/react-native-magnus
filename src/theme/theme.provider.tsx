import * as React from "react";

import { ThemeType } from "./type";

import { ThemeContext } from "./theme.context";
import { useTheme } from "./theme.hook";

export interface ThemeProviderProps {
  theme?: ThemeType;
  children?: React.ReactNode;
}

export const ThemeProvider: React.FunctionComponent<ThemeProviderProps> = (
  props: ThemeProviderProps
) => {
  const { theme: themeProp, children } = props;
  const theme = useTheme(themeProp);

  return (
    <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
  );
};
