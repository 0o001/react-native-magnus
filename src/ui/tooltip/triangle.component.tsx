import * as React from 'react';

import { View as RNView, StyleSheet } from 'react-native';
import { TriangelProps } from './tooltip.type';

const Triangle: React.FunctionComponent<TriangelProps> = (props) => {
  const { style, invert } = props;
  let computedStyle = style;

  if (invert) {
    computedStyle = StyleSheet.flatten([
      style,
      {
        transform: [{ rotate: '180deg' }],
      },
    ]);
  }

  return <RNView style={computedStyle} />;
};

Triangle.defaultProps = {
  invert: false,
};

export { Triangle };
