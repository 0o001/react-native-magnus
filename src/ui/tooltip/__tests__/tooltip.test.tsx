import React from "react";
import renderer from "react-test-renderer";

import { Tooltip } from "../tooltip.component";
import { ThemeProvider } from "../../../theme";

describe("Div Component", () => {
  it("should render without issues", () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Tooltip text="Tooltip text">Tooltip</Tooltip>
      </ThemeProvider>
    );

    expect(instance.toJSON()).toMatchSnapshot();

    const typeInst = instance.root.findByType(Tooltip);
    expect(typeInst.props.children).toBe("Tooltip");
  });
});
