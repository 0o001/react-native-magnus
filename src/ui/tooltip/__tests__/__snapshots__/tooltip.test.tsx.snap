// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Div Component should render without issues 1`] = `
<View>
  <View>
    Tooltip
  </View>
  <Modal
    hardwareAccelerated={false}
    onDismiss={[Function]}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      accessible={false}
      clickable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "transparent",
            "borderRadius": 4,
            "marginTop": -8,
            "opacity": 0,
            "position": "absolute",
            "top": 8,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "borderBottomColor": "#262626",
              "borderBottomWidth": 10,
              "borderLeftColor": "transparent",
              "borderLeftWidth": 8,
              "borderRightColor": "transparent",
              "borderRightWidth": 8,
              "borderStyle": "solid",
              "height": 0,
              "left": -8,
              "top": 0,
              "width": 0,
            }
          }
        />
        <View
          style={
            Object {
              "alignItems": undefined,
              "alignSelf": "center",
              "backgroundColor": "#262626",
              "borderRadius": 12,
              "flexDirection": undefined,
              "flexWrap": undefined,
              "justifyContent": undefined,
              "marginHorizontal": 12,
              "padding": 12,
            }
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#FFFFFF",
                "fontSize": 15,
                "lineHeight": 22.5,
                "textAlign": "auto",
                "textDecorationColor": "transparent",
                "textTransform": "none",
              }
            }
          >
            Tooltip text
          </Text>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
